src/algorithms/StatePreparation.cpp

```diff
@@ -0,0 +1,285 @@
+#include <cstdint>
+#include <iterator>
+#include <numeric>
+#include <tuple>
+#include <vector>
+
+static const double EPS = 1e-10;
+
+namespace qc {
+using Matrix = std::vector<std::vector<double>>;
+
+template <typename T> [[nodiscard]] auto twoNorm(std::vector<T> vec) -> double {
+  double norm = 0;
+  for (auto elem : vec) {
+    norm += std::norm(elem);
+  }
+  return sqrt(norm);
+}
+
+template <typename T>
+[[nodiscard]] auto isNormalized(std::vector<T> vec) -> bool {
+  return std::abs(1 - twoNorm(vec)) < EPS;
+}
+
+[[nodiscard]] auto kroneckerProduct(Matrix matrixA, Matrix matrixB) -> Matrix {
+  size_t const rowA = matrixA.size();
+  size_t const rowB = matrixB.size();
+  size_t const colA = matrixA[0].size();
+  size_t const colB = matrixB[0].size();
+  // initialize size
+  Matrix newMatrix{(rowA * rowB), std::vector<double>(colA * colB, 0)};
+  // code taken from RosettaCode slightly adapted
+  for (size_t i = 0; i < rowA; ++i) {
+    // k loops till rowB
+    for (size_t j = 0; j < colA; ++j) {
+      // j loops till colA
+      for (size_t k = 0; k < rowB; ++k) {
+        // l loops till colB
+        for (size_t l = 0; l < colB; ++l) {
+          // Each element of matrix A is
+          // multiplied by whole Matrix B
+          // resp and stored as Matrix C
+          newMatrix[i * rowB + k][j * colB + l] = matrixA[i][j] * matrixB[k][l];
+        }
+      }
+    }
+  }
+  return newMatrix;
+}
+
+[[nodiscard]] auto createIdentity(size_t size) -> Matrix {
+  Matrix identity{
+      std::vector<std::vector<double>>(size, std::vector<double>(size, 0))};
+  for (size_t i = 0; i < size; ++i) {
+    identity[i][i] = 1;
+  }
+  return identity;
+}
+
+[[nodiscard]] auto matrixVectorProd(const Matrix& matrix,
+                                    std::vector<double> vector)
+    -> std::vector<double> {
+  std::vector<double> result;
+  for (const auto& matrixVec : matrix) {
+    double sum{0};
+    for (size_t i = 0; i < matrixVec.size(); ++i) {
+      sum += matrixVec[i] * vector[i];
+    }
+    result.push_back(sum);
+  }
+  return result;
+}
+
+// recursive implementation that returns multiplexer circuit
+/**
+ * @param target_gate : Ry or Rz gate to apply to target qubit, multiplexed
+ * over all other "select" qubits
+ * @param angles : list of rotation angles to apply Ry and Rz
+ * @param lastCnot : add last cnot if true
+ * @return multiplexer circuit as QuantumComputation
+ */
+[[nodiscard]] auto multiplex(OpType targetGate, std::vector<double> angles,
+                             bool lastCnot) -> QuantumComputation {
+  size_t const listLen = angles.size();
+  double const localNumQubits =
+      std::floor(std::log2(static_cast<double>(listLen))) + 1;
+  QuantumComputation multiplexer{static_cast<size_t>(localNumQubits)};
+  // recursion base case
+  if (localNumQubits == 1) {
+    multiplexer.emplace_back<StandardOperation>(Controls{}, 0, targetGate,
+                                                std::vector{angles[0]});
```

Comment:
```suggestion
    multiplexer.emplace_back<StandardOperation>(0, targetGate,
                                                std::vector{angles[0]});
```
should work just fine

Reply:
Done in commit b3209bb